Source/WebKit/UIProcess/API/glib/WebKitNotificationProvider.cpp

```diff
@@ -80,69 +81,95 @@ WebKitNotificationProvider::WebKitNotificationProvider(WebNotificationManagerPro
 
 WebKitNotificationProvider::~WebKitNotificationProvider()
 {
+    if (m_observerRegistered)
+        NotificationService::singleton().removeObserver(*this);
+
     m_notificationManager->setProvider(nullptr);
 }
 
-void WebKitNotificationProvider::notificationCloseCallback(WebKitNotification* notification, WebKitNotificationProvider* provider)
+void WebKitNotificationProvider::apiNotificationCloseCallback(WebKitNotification* notification, WebKitNotificationProvider* provider)
 {
     uint64_t notificationID = webkit_notification_get_id(notification);
     Vector<RefPtr<API::Object>> arrayIDs;
     arrayIDs.append(API::UInt64::create(notificationID));
     provider->m_notificationManager->providerDidCloseNotifications(API::Array::create(WTFMove(arrayIDs)).ptr());
-    provider->m_notifications.remove(notificationID);
+    provider->m_apiNotifications.remove(notificationID);
 }
 
-void WebKitNotificationProvider::notificationClickedCallback(WebKitNotification* notification, WebKitNotificationProvider* provider)
+void WebKitNotificationProvider::apiNotificationClickedCallback(WebKitNotification* notification, WebKitNotificationProvider* provider)
 {
     provider->m_notificationManager->providerDidClickNotification(webkit_notification_get_id(notification));
 }
 
-void WebKitNotificationProvider::withdrawAnyPreviousNotificationMatchingTag(const CString& tag)
+void WebKitNotificationProvider::closeAPINotification(uint64_t notificationID)
+{
+    if (auto notification = m_apiNotifications.take(notificationID))
+        webkit_notification_close(notification.get());
+}
+
+void WebKitNotificationProvider::withdrawAnyPreviousAPINotificationMatchingTag(const CString& tag)
 {
     if (!tag.length())
         return;
 
-    for (auto& notification : m_notifications.values()) {
+    for (auto& notification : m_apiNotifications.values()) {
         if (tag == webkit_notification_get_tag(notification.get())) {
-            webkit_notification_close(notification.get());
+            closeAPINotification(webkit_notification_get_id(notification.get()));
             break;
         }
     }
 
 #ifndef NDEBUG
```

Comment:
Good opportunity for a drive-by fix:

```suggestion
#if ASSERT_ENABLED
```

(`NDEBUG` controls glibc assertions, not WebKit assertions.)